### PR TITLE
Sc 406 fix and update fpgaflow to align with other flows

### PR DESCRIPTION
**Pull request by petergrossmann21**
This is a "hello world" demonstrator only.  The key features are:  separate flows and targets for VPR/nextpnr based flows as the execution models of each tool are fairly different; tests that exercise each flow with tiny circuits and tiny architecture models.

### CODE REVIEW

File: examples/fpga_flow/designs/adder/adder.vhd [hunk 1]
```diff
@@ -0,0 +1,33 @@
+--adder.vhd
+--Peter Grossmann
+--24 May 2023
+--$Id$
+--$Log$
```
Comment (gadfort) [inline examples/fpga_flow/designs/adder/adder.vhd:5]: remove (same for all)

File: siliconcompiler/flows/fpga_nextpnr_flow.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
+import siliconcompiler
```
Comment (gadfort) [inline siliconcompiler/flows/fpga_nextpnr_flow.py:1]: why not make this and fpga_vpr_flow a single flow with an input to indicate VPR vs NEXTPNR?

File: siliconcompiler/flows/fpga_nextpnr_flow.py [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+from siliconcompiler.tools.yosys import syn_fpga
+from siliconcompiler.tools.nextpnr import apr as nextpnr_apr
+
+
+############################################################################
+# DOCS
+############################################################################
+def make_docs(chip):
+    return setup(chip)
+
```
Comment (gadfort) [inline siliconcompiler/flows/fpga_nextpnr_flow.py:14]: remove

File: siliconcompiler/flows/fpga_nextpnr_flow.py [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+############################################################################
+# Flowgraph Setup
+############################################################################
+def setup(chip, flowname='fpga_nextpnr_flow'):
+    '''
+    A configurable FPGA compilation flow.
+
+    The 'fpgaflow' module is a configurable FPGA flow with support for
+    open source and commercial tool flows.
```
Comment (gadfort) [inline siliconcompiler/flows/fpga_nextpnr_flow.py:24]: which commercial tools? I would cleanup this docstring

File: siliconcompiler/flows/fpga_nextpnr_flow.py [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+    * **import**: Sources are collected and packaged for compilation
+    * **syn**: Synthesize RTL into an device specific netlist
+    * **place**: FPGA specific placement and routing step
+    * **route**: FPGA specific placement and routing step
+    * **bitstream**: Bitstream generation
+
+    The fpgaflow can be configured througthe following schema parameters
+
+    Schema keypaths:
+
```
Comment (gadfort) [inline siliconcompiler/flows/fpga_nextpnr_flow.py:38]: ??

File: siliconcompiler/flows/fpga_nextpnr_flow.py [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+        flowtools.append((step, tool_modules[step]))
+
+    # Minimal setup
+    index = '0'
+    prevstep = None
+    for step, tool_module in flowtools:
+        # Flow
+        flow.node(flowname, step, tool_module)
+        if prevstep:
+            flow.edge(flowname, prevstep, step)
```
Comment (gadfort) [inline siliconcompiler/flows/fpga_nextpnr_flow.py:62]: maybe just use `flow.pipe` instead?

File: siliconcompiler/flows/fpga_vpr_flow.py [hunk 7]
```diff
@@ -0,0 +1,84 @@
+import siliconcompiler
```
Comment (gadfort) [inline siliconcompiler/flows/fpga_vpr_flow.py:1]: merge with the nextpnr one

File: siliconcompiler/targets/fpga_nextpnr_flow_demo.py [hunk 8]
```diff
@@ -11,24 +11,24 @@ def make_docs(chip):
 ####################################################
 # Target Setup
 ####################################################
-def setup(chip):
+def setup(chip, toolflow='vpr'):
```
Comment (gadfort) [inline siliconcompiler/targets/fpga_nextpnr_flow_demo.py:14]: you don't use `toolflow`, please return this back to `fpga_flow_demo` and use that flag to indicate which flow you are loading

File: siliconcompiler/tools/genfasm/bitstream.py [hunk 9]
```diff
@@ -37,26 +32,27 @@ def setup(chip):
...
-    tool = "genfasm"
-
-    chip.add('tool', tool, 'task', task, 'option', [f"--route_chan_width {find_chann_width()}"],
-             step=step, index=index)
+# def pre_process(chip):
 
+    # step = chip.get('arg', 'step')
+    # index = chip.get('arg', 'index')
+    # task = chip._get_task(step, index)
+    # tool = "genfasm"
```
Comment (gadfort) [inline siliconcompiler/tools/genfasm/bitstream.py:41]: remove if not needed

File: siliconcompiler/tools/vpr/route.py [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+    chip.set('tool', tool, 'exe', tool, clobber=False)
+    chip.set('tool', tool, 'version', '0.0', clobber=False)
+
+    chip.set('tool', tool, 'task', task, 'threads', os.cpu_count(),
+             step=step, index=index, clobber=False)
+
+    # TO-DO: PRIOROTIZE the post-routing packing results?
+    design = chip.top()
+    chip.add('tool', tool, 'task', task, 'output', design + '.route', step=step, index=index)
+    chip.add('tool', tool, 'task', task, 'output', 'vpr_stdout.log', step=step, index=index)
```
Comment (gadfort) [inline siliconcompiler/tools/vpr/route.py:27]: is this a real output?

File: siliconcompiler/tools/vpr/route.py [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+
+    chip.set('tool', tool, 'exe', tool, clobber=False)
+    chip.set('tool', tool, 'version', '0.0', clobber=False)
+
+    chip.set('tool', tool, 'task', task, 'threads', os.cpu_count(),
+             step=step, index=index, clobber=False)
+
+    # TO-DO: PRIOROTIZE the post-routing packing results?
+    design = chip.top()
+    chip.add('tool', tool, 'task', task, 'output', design + '.route', step=step, index=index)
```
Comment (gadfort) [inline siliconcompiler/tools/vpr/route.py:26]: use `'f{design}.route'`

File: siliconcompiler/tools/vpr/route.py [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+
+def pre_process(chip):
+
+    # have to rename the net connected to unhooked pins from $undef to unconn
+    # as VPR uses unconn keywords to identify unconnected pins
+
+    step = chip.get('arg', 'step')
+    index = chip.get('arg', 'index')
+    design = chip.top()
+    blif_file = f"{chip._getworkdir()}/{step}/{index}/inputs/{design}.blif"
```
Comment (gadfort) [inline siliconcompiler/tools/vpr/route.py:54]: you can just do `inputs/{}.blif`

File: siliconcompiler/tools/vpr/route.py [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+    print(blif_file)
+    with open(blif_file, 'r+') as f:
+        netlist = f.read()
+        f.seek(0)
+        netlist = re.sub(r'\$undef', 'unconn', netlist)
+        f.write(netlist)
+        f.truncate()
```
Comment (gadfort) [inline siliconcompiler/tools/vpr/route.py:61]: don't overwrite the input, make it a different file so we don't modify the inputs.

File: siliconcompiler/tools/vpr/route.py [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+    index = chip.get('arg', 'index')
+    task = chip._get_task(step, index)
+
+    for file in chip.get('tool', 'vpr', 'task', task, 'output', step=step, index=index):
+        shutil.copy(file, 'outputs')
+    design = chip.top()
+    # Forward all of the prior step inputs forward for bitstream generation
+    shutil.copy(f'inputs/{design}.blif', 'outputs')
+    shutil.copy(f'inputs/{design}.net', 'outputs')
+    shutil.copy(f'inputs/{design}.place', 'outputs')
```
Comment (gadfort) [inline siliconcompiler/tools/vpr/route.py:82]: why? shouldn't this be handled in the graph setup instead?

File: siliconcompiler/tools/vpr/vpr.py [hunk 15]
```diff
@@ -26,6 +26,48 @@ def make_docs(chip):
...
+    step = chip.get('arg', 'step')
+    index = chip.get('arg', 'index')
+    # task = chip._get_task(step, index)
+
+    options = []
+
+    topmodule = chip.top()
+    blif = f"inputs/{topmodule}.blif"
+
+    for arch in chip.get('fpga', 'arch'):
```
Comment (gadfort) [inline siliconcompiler/tools/vpr/vpr.py:40]: I don't see fpga/arch as required.

File: siliconcompiler/tools/vpr/vpr.py [hunk 16]
```diff
@@ -26,6 +26,48 @@ def make_docs(chip):
...
+        options.append(arch)
+
+    options.append(blif)
+
+    if 'sdc' in chip.getkeys('input'):
+        options.append(f"--sdc_file {chip.get('input', 'fpga', 'sdc', step=step, index=index)}")
```
Comment (gadfort) [inline siliconcompiler/tools/vpr/vpr.py:46]: always false. you need to use find_files for this but you cannot do this in `setup()` it must only be called in `pre_process`

File: siliconcompiler/tools/yosys/syn_fpga.tcl [hunk 17]
```diff
@@ -4,19 +4,159 @@ set job_name [dict get $sc_cfg option jobname]
 set step [dict get $sc_cfg arg step]
 set index [dict get $sc_cfg arg index]
 
+set sc_memmap_techlib "None"
+set sc_mem_techlib "None"
+set sc_lut_techlib "None"
+set sc_dsp_techlib "None"
+set sc_flop_techlib "None"
```
Comment (gadfort) [inline siliconcompiler/tools/yosys/syn_fpga.tcl:11]: why are these here?

File: siliconcompiler/tools/yosys/syn_fpga.tcl [hunk 18]
```diff
@@ -4,19 +4,159 @@ set job_name [dict get $sc_cfg option jobname]
...
+set sc_syn_lut_size [dict get $sc_cfg tool $sc_tool task $sc_step var lut_size ]
```
Comment (gadfort) [inline siliconcompiler/tools/yosys/syn_fpga.tcl:12]: no space at the end.

File: siliconcompiler/tools/yosys/syn_fpga.tcl [hunk 19]
```diff
@@ -4,19 +4,159 @@ set job_name [dict get $sc_cfg option jobname]
...
+#          synthesis command.  It will be necessary to do some vetting and
+#          decide what to really support and what to omit based on the yosys
+#          support (or the place and route support) being too skimpy.
+#          -PG 1/13/2023
+
 if {[string match {ice*} $sc_partname]} {
     yosys synth_ice40 -top $sc_design -json "${sc_design}_netlist.json"
 } else {
-    yosys script "${build_dir}/${sc_design}/${job_name}/${step}/${index}/inputs/vpr_yosyslib/synthesis.ys"
+    #yosys script "${build_dir}/${sc_design}/${job_name}/${step}/${index}/inputs/vpr_yosyslib/synthesis.ys"
```
Comment (gadfort) [inline siliconcompiler/tools/yosys/syn_fpga.tcl:41]: if not needed, remove

File: siliconcompiler/tools/yosys/syn_fpga.tcl [hunk 20]
```diff
@@ -4,19 +4,159 @@ set job_name [dict get $sc_cfg option jobname]
...
+    #Match VPR reference flow's hierarchy check, including their comments
```
Comment (gadfort) [inline siliconcompiler/tools/yosys/syn_fpga.tcl:42]: you should provide the link to the source.

File: siliconcompiler/tools/yosys/syn_fpga.tcl [hunk 21]
```diff
@@ -4,19 +4,159 @@ set job_name [dict get $sc_cfg option jobname]
...
+    # about, such as hand-instantiated (not inferred) memories
+    #***NOTE:  Removed -libdir as it's nonsensical with the
+    #          use case of this script -PG 1/11/2023
+    yosys hierarchy -check -auto-top
+
+    #Rename top level module to match selected design name;
+    #This allows us to enforce some naming consistency in
+    #automated flows
+    #***NOTE:  because rename is a Tcl keyword the yosys -import
+    #          trick doesn't work
```
Comment (gadfort) [inline siliconcompiler/tools/yosys/syn_fpga.tcl:58]: you can remove this, our preference is to not rely on the import methos anyways

File: siliconcompiler/tools/yosys/syn_fpga.tcl [hunk 22]
```diff
@@ -4,19 +4,159 @@ set job_name [dict get $sc_cfg option jobname]
...
+    yosys rename -top ${sc_design}
+
+    yosys proc
+    #select -module ${sc_design}
+    #flatten -wb
+    #flatten
```
Comment (gadfort) [inline siliconcompiler/tools/yosys/syn_fpga.tcl:64]: remove

File: siliconcompiler/tools/yosys/syn_fpga.tcl [hunk 23]
```diff
@@ -4,19 +4,159 @@ set job_name [dict get $sc_cfg option jobname]
...
+    yosys opt
+    yosys wreduce
+    yosys peepopt
+    yosys opt_clean
+    yosys share
+    yosys opt
+    yosys memory -nomap
+    yosys opt -full
+
+    #Do our own thing from here
```
Comment (gadfort) [inline siliconcompiler/tools/yosys/syn_fpga.tcl:81]: what does this mean?

File: siliconcompiler/tools/yosys/syn_fpga.tcl [hunk 24]
```diff
@@ -4,19 +4,159 @@ set job_name [dict get $sc_cfg option jobname]
...
+    yosys techmap -map +/techmap.v
+
+    #***TO DO:  Figure out how to deal with this in an SC-compatible way
```
Comment (gadfort) [inline siliconcompiler/tools/yosys/syn_fpga.tcl:84]: lets discuss

File: siliconcompiler/tools/yosys/syn_fpga.tcl [hunk 25]
```diff
@@ -4,19 +4,159 @@ set job_name [dict get $sc_cfg option jobname]
...
+    if { $sc_flop_techlib eq "None" } {
+	yosys dfflegalize -cell \$_DFF_P_ 01 -cell \$_DFF_P??_ 01 
+    }
+
+    #opt
+    #opt_clean
```
Comment (gadfort) [inline siliconcompiler/tools/yosys/syn_fpga.tcl:90]: remove

File: siliconcompiler/tools/yosys/syn_fpga.tcl [hunk 26]
```diff
@@ -4,19 +4,159 @@ set job_name [dict get $sc_cfg option jobname]
...
+
+    #Perform preliminary buffer insertion before passing to ABC to help reduce
+    #the overhead of final buffer insertion downstream
+    yosys insbuf
+
+    yosys abc -lut $sc_syn_lut_size -dff
+    #abc -script $env(FPGA_ARCHITECT_ROOT)/test/scripts/fpgaa_abc_flow.scr
```
Comment (gadfort) [inline siliconcompiler/tools/yosys/syn_fpga.tcl:97]: remove? same for all commented out commands

File: siliconcompiler/tools/yosys/syn_fpga.tcl [hunk 27]
```diff
@@ -4,19 +4,159 @@ set job_name [dict get $sc_cfg option jobname]
...
+    #size with that step
+    #opt
+    #clean
+
+    yosys flatten
+    yosys clean
+
+    #***NOTE:  If you flatten after this step instead of before Yosys barfs,
+    #          would be nice to have a deeper understanding of why that is
+    #          -PG 1/16/2023
```
Comment (gadfort) [inline siliconcompiler/tools/yosys/syn_fpga.tcl:112]: whats the issue on github for this in yosys repo?

File: siliconcompiler/tools/yosys/syn_fpga.tcl [hunk 28]
```diff
@@ -4,19 +4,159 @@ set job_name [dict get $sc_cfg option jobname]
...
+    set sc_memmap_techlib [ dict get $sc_cfg tool $sc_tool task $sc_step var memmap ]
+    if { $sc_memmap_techlib ne "None" } {
+	yosys memory_libmap -lib $sc_memmap_techlib
+	#***HACK:  Eliminate $reduce_or with this mapping here
+	#          until we have a cleaner solution
```
Comment (gadfort) [inline siliconcompiler/tools/yosys/syn_fpga.tcl:117]: why is this needed?

File: siliconcompiler/tools/yosys/syn_fpga.tcl [hunk 29]
```diff
@@ -4,19 +4,159 @@ set job_name [dict get $sc_cfg option jobname]
...
+    if [dict exists $sc_cfg tool $sc_tool var $sc_step $sc_index techmap ] {
+	set sc_techmap [ dict get $sc_cfg tool $sc_tool var $sc_step $sc_index techmap ]
+	# techmap_defines = [ dict get $sc_cfg tool $sc_tool var $sc_step $sc_index techmap_define ]
+	# techmap_define_args = ''
+	# foreach techmap_define $techmap_defines {
+	#    set techmap_define_args "$techmap_define_args -D$techmap_define"
+	# }
+	foreach mapfile $sc_techmap {
+	    # Add $techmap_define_args when ready
+	    yosys techmap -map $mapfile
```
Comment (gadfort) [inline siliconcompiler/tools/yosys/syn_fpga.tcl:145]: after techmap you need to rerun `yosys techmap` to ensure everything is mapped.

File: siliconcompiler/tools/yosys/syn_fpga.tcl [hunk 30]
```diff
@@ -4,19 +4,159 @@ set job_name [dict get $sc_cfg option jobname]
...
+	    yosys techmap -map $mapfile 
+	}
+    }
+
+    #***NOTE:  Can't do a hierarchy check here as the techmap results
+    #          get flagged
+    #***TO DO:  Review why this might have been in the flow during
+    #           previous development phases and revisit what if anything to
+    #           do going forward -PG 12/23/2022
+    #hierarchy -check -purge_lib
```
Comment (gadfort) [inline siliconcompiler/tools/yosys/syn_fpga.tcl:154]: what is the issue here?

File: siliconcompiler/tools/yosys/syn_fpga.tcl [hunk 31]
```diff
@@ -4,19 +4,159 @@ set job_name [dict get $sc_cfg option jobname]
...
+
+    yosys clean
+
 }
 
-# turn off echo to prevent the stat command from showing up in the json file
 yosys echo off
-yosys tee -o ./reports/stat.json stat -json -top $sc_design
+yosys tee -o ./reports/stat.json stat -json -top $sc_design
```
Comment (gadfort) [inline siliconcompiler/tools/yosys/syn_fpga.tcl:161]: space at the end of the line, please remove

**Review by gadfort — CHANGES_REQUESTED** [review]
Lets discuss some of these, I did not review the unit tests, yet

File: siliconcompiler/tools/yosys/syn_fpga.tcl (hunk 28 above)
Comment (petergrossmann21) [reply siliconcompiler/tools/yosys/syn_fpga.tcl:117]: This synthesis recipe was brought in from an FPGA Architect setup that had a few outstanding details to iron out.  The $reduce_or yosys primitive handling is an open question for me regarding technology mapping for FPGAs that have their own technology map libraries.

File: siliconcompiler/tools/yosys/syn_fpga.tcl (hunk 30 above)
Comment (petergrossmann21) [reply siliconcompiler/tools/yosys/syn_fpga.tcl:154]: I will need to reproduce the error messages referenced by this note to comment further.

File: siliconcompiler/flows/fpga_nextpnr_flow.py (hunk 2 above)
Comment (petergrossmann21) [reply siliconcompiler/flows/fpga_nextpnr_flow.py:1]: nextpnr does not appear to have the ability to run placement, routing, and bitstream generation as separate steps.  It seemed clunky to have FPGA targets and flows for two different place and route tools when they wouldn't share a flowgraph or anything of substance besides their synthesis front end and the users were going to have to pass the tool selection in as a load_target() argument anyway.  That said, I could be talked in to re-unifying the flow with some additional pointers on the methodology for if-elsing the flow graph based on a tool selection.

File: siliconcompiler/tools/yosys/syn_fpga.tcl (hunk 31 above)
Comment (petergrossmann21) [reply siliconcompiler/tools/yosys/syn_fpga.tcl:161]: Done.

File: siliconcompiler/tools/yosys/syn_fpga.tcl (hunk 27 above)
Comment (petergrossmann21) [reply siliconcompiler/tools/yosys/syn_fpga.tcl:112]: I have not filed a github issue for this at YosysHQ; nor am I aware of whether someone else has.

File: siliconcompiler/flows/fpga_nextpnr_flow.py (hunk 4 above)
Comment (petergrossmann21) [reply siliconcompiler/flows/fpga_nextpnr_flow.py:24]: Comment deleted.

File: examples/fpga_flow/designs/adder/adder.vhd (hunk 1 above)
Comment (petergrossmann21) [reply examples/fpga_flow/designs/adder/adder.vhd:5]: Done.

File: siliconcompiler/flows/fpga_nextpnr_flow.py (hunk 2 above)
Comment (petergrossmann21) [reply siliconcompiler/flows/fpga_nextpnr_flow.py]: Flows have been consolidated.  ICE40 and ZeroASIC demo architecture are offered as distinct targets with subtly different settings.
